### PR TITLE
Add support for breaking out of the sandbox when needed

### PR DESCRIPTION
**Pull request by orta**
We have a link that needs to escape the eigen sandboxing around routing, in this case - the link is a link that safari can handle and create an event in someone's calendar. In the future, well, it could be anything. 

So we want to provide a way to optionally skip this routing for _non_ artsy URLs. So, prerequisites are: not in this list `artsyHosts = [NSSet setWithObjects:@"art.sy", @"artsyapi.com", @"artsy.net", @"m.artsy.net", @"staging.artsy.net", @"m-staging.artsy.net", nil];`, and has a query string which includes the string `eigen_escape_sandbox`. 

/cc @broskoski

### CODE REVIEW

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%">1 Warning</th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td><h4>Found unstubbbed networking requests</h4>

<ul>
<li><a href="https://github.com/search?q=ARArtworkViewControllerBuyButton+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARArtworkViewControllerBuyButtonSpec</a> in <code>test_buy_button__posts_order_if_artwork_has_no_edition_sets</code></li>
<li><a href="https://github.com/search?q=ARCreateAccountViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARCreateAccountViewControllerSpec</a> in <code>test_warning_view__displays_a_message</code></li>
<li><a href="https://github.com/search?q=ARCreateAccountViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARCreateAccountViewControllerSpec</a> in <code>test_warning_view__displays_a_message</code></li>
<li><a href="https://github.com/search?q=AREmbeddedModelPreviewViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">AREmbeddedModelPreviewViewControllerSpec</a> in <code>test_iPhone_peek__shows_favorite_and_follow_buttons_for_logged_in_user</code></li>
<li><a href="https://github.com/search?q=AREmbeddedModelPreviewViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">AREmbeddedModelPreviewViewControllerSpec</a> in <code>test_iPhone_peek__shows_favorite_and_follow_buttons_for_logged_in_user</code></li>
<li><a href="https://github.com/search?q=AREmbeddedModelPreviewViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">AREmbeddedModelPreviewViewControllerSpec</a> in <code>test_iPhone_peek__passes_artwork_to_share_controller</code></li>
<li><a href="https://github.com/search?q=ARFairViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARFairViewControllerSpec</a> in <code>test_creates_an_NSUserActivity</code></li>
<li><a href="https://github.com/search?q=ARGeneViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARGeneViewControllerSpec</a> in <code>test_creates_an_NSUserActivity</code></li>
<li><a href="https://github.com/search?q=ARGeneViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARGeneViewControllerSpec</a> in <code>test_creates_an_NSUserActivity</code></li>
<li><a href="https://github.com/search?q=ARTopMenuViewController+repo%3Aartsy%2Feigen&amp;ref=searchresults&amp;type=Code&amp;utf8=%E2%9C%93">ARTopMenuViewControllerSpec</a> in <code>test_navigation__favorites__logged_in__tab_behavior__already_selected__pops_to_root</code></li>
</ul>
</td>
    </tr>
    
  </tbody>
</table>
  

  


<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/KrauseFx/danger/">danger</a>
</p>

**Comment by broskoski** [comment]
:clap: :clap: :clap: :+1: :confetti_ball: :cake:

File: Artsy_Tests/App_Tests/ARSwitchBoardTests.m [hunk 1]
```diff
@@ -26,6 +26,7 @@ - (NSURL *)resolveRelativeUrl:(NSString *)path;
 - (id)routeInternalURL:(NSURL *)url fair:(Fair *)fair;
 - (void)openURLInExternalService:(NSURL *)url;
 - (void)updateRoutes;
+@property (nonatomic, strong) UIApplication *sharedApplication;
```
Comment (alloy) [inline Artsy_Tests/App_Tests/ARSwitchBoardTests.m]: What is this used for?

File: Artsy/App/ARSwitchBoard.m [hunk 2]
```diff
@@ -335,4 +338,11 @@ - (NSURL *)fixHostForURL:(NSURL *)url
     return url;
 }
 
+#pragma mark DI -
+
+- (UIApplication *)sharedApplication
+{
+    return _sharedApplication ?: [UIApplication sharedApplication];
+}
```
Comment (alloy) [inline Artsy/App/ARSwitchBoard.m]: I don’t see where this is used?

File: Artsy_Tests/App_Tests/ARSwitchBoardTests.m (hunk 1 above)
Comment (orta) [inline Artsy_Tests/App_Tests/ARSwitchBoardTests.m]: nothing ATM, but it's better to have it support DI here - the tests should be using it, but given all of them use mocks - I figured I'd leave it in

File: Artsy/App/ARSwitchBoard.m [hunk 3]
```diff
@@ -267,13 +268,15 @@ - (UIViewController *)loadURL:(NSURL *)url fair:(Fair *)fair
         return [self routeInternalURL:fixedURL fair:fair];
 
     } else if ([ARRouter isWebURL:url]) {
-        /// Is is a webpage we could open in webkit?
-        if (ARIsRunningInDemoMode) {
-            [[UIApplication sharedApplication] openURL:url];
+        /// Is is a webpage we could open in webkit?, or need to break out to safari (see PR #1195)
+        if (ARIsRunningInDemoMode || [url.query containsString:@"eigen_escape_sandbox"]) {
```
Comment (alloy) [inline Artsy/App/ARSwitchBoard.m]: Maybe move this to an app constant?

Comment (orta) [inline Artsy/App/ARSwitchBoard.m]: I like that

Comment (orta) [inline Artsy/App/ARSwitchBoard.m]: will do it after my current work

File: Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m [hunk 4]
```diff
@@ -158,7 +158,10 @@ - (WKNavigationActionPolicy)shouldLoadNavigationAction:(WKNavigationAction *)nav
         NSURL *URL = navigationAction.request.URL;
         ARSwitchBoard *switchboard = [ARSwitchBoard sharedInstance];
         if ([switchboard canRouteURL:URL]) {
-            [switchboard presentViewController:[switchboard loadURL:URL]];
+            UIViewController *controller = [switchboard loadURL:URL];
+            if (controller) {
+                [switchboard presentViewController:controller];
+            }
```
Comment (orta) [inline Artsy/View_Controllers/Web_Browsing/ARExternalWebBrowserViewController.m]: this was the only case I could find where we were not taking the nullability into account,

**Comment by orta** [comment]
updated :+1:

**Pull request merged**